sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (C) 2020 chrysn
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @{
+ *
+ * @file
+ * @brief       CoAP file server implementation
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "checksum/fletcher32.h"
+#include "net/gcoap/fileserver.h"
+#include "net/gcoap.h"
+#include "vfs.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/** Maximum length of an expressible path, including the trailing 0 character. */
+#define COAPFILESERVER_PATH_MAX (64)
+
+/** Data extracted from a request on a file */
+struct requestdata {
+    /** 0-terminated expanded file name in the VFS */
+    char namebuf[COAPFILESERVER_PATH_MAX];
+    uint32_t blocknum2;
+    unsigned int szx2; /* would prefer uint8_t but that's what coap_get_blockopt gives */
+    uint32_t etag;
+    bool etag_sent;
+};
+
+/**
+ * @brief  Return true if path/name is a directory.
+ */
+static bool entry_is_dir(char *path, const char *name)
+{
+    struct stat stat;
+    size_t path_len = strlen(path);
+    size_t name_len = strlen(name);
+
+    if (path_len + name_len + 1 >= COAPFILESERVER_PATH_MAX) {
+        return false;
+    }
+
+    /* re-use path buffer, it is already COAPFILESERVER_PATH_MAX bytes long */
+    path[path_len] = '/';
+    path[path_len + 1] = 0;
+    strncat(path, name, COAPFILESERVER_PATH_MAX - 1);
```

Comment:
Why not `strncpy`, since `path_len` already contains the current length of path?

Beware: I think that strncpy and strncat have a stupid API, as they may leave the destination path without terminating zero byte if the destination buffer is not large enough.

Reply:
Do you mean

```C
    strncpy(path + path_len, name, COAPFILESERVER_PATH_MAX - path_len - 1);
```

That looks a bit more fragile to me, but might be more efficient.

Reply:
```C
strncpy(path + path_len + 1, name, COAPFILESERVER_PATH_MAX - path_len - 2);
```
+1 because of '/' and -2 also for the '/' and a '\0'.

Reply:
I am fine with both. @maribu, choose readability over efficiency, or not ?